openfpga/src/fpga_bitstream/write_text_fabric_bitstream.cpp

```diff
@@ -245,6 +245,142 @@ static int write_memory_bank_flatten_fabric_bitstream_to_text_file(
   return status;
 }
 
+/********************************************************************
+ * Write the fabric bitstream fitting a memory bank protocol
+ * to a plain text file in efficient method
+ *
+ * Old function is write_memory_bank_flatten_fabric_bitstream_to_text_file()
+ *
+ * As compared to original function, based on 100K LE FPGA:
+ *  1. Original function used 600 seconds and need 80G Bytes of memory
+ *  2. This new function only needs 1 second and 4M Bytes
+ *
+ * Old function only print WL in decremental order. It is not by intentional
+ * It is because of the map-key ordering
+ * In QL Memory Bank with Flatten BL/WL, data is stored by WL address,
+ *   where we use WL string as map key
+ *     WL #0 --- "1000000000000 .... 0000"
+ *     WL #1 --- "0100000000000 .... 0000"
+ *     WL #n-1 --- "0000000000000 .... 0001
+ * From string comparison wise, WL #n-1 will be first, and WL #0 will be last
+ * The sequence of WL does not really matter,  but preferrable in some ordering
+ *   manner. Using map key as ordering cannot guarantee the determinstic
+ *
+ * This new way of writting fabric guarantee the WL order in 100% deterministc
+ *   way: either incremental (default) or decremental
+ *
+ * Return:
+ *  - 0 if succeed
+ *  - 1 if critical errors occured
+ *******************************************************************/
+static int fast_write_memory_bank_flatten_fabric_bitstream_to_text_file(
+  std::fstream& fp, const bool& fast_configuration,
+  const bool& bit_value_to_skip, const FabricBitstream& fabric_bitstream,
+  const bool& keep_dont_care_bits, const bool& wl_incremental_order) {
+  int status = 0;
+
+  std::string dont_care_bit = "0";
+  if (keep_dont_care_bits) {
+    dont_care_bit = "x";
+  }
+  const FabricBitstreamMemoryBank* memory_bank =
```

Comment:
It is o.k. to have sub data structure inside the FabricBitstream. However, when designing the data structure, we want to make the FabricBitstream data structure to be the one and the only one that client functions should acquire information. Therefore, I suggest to avoid use methods/APIs from the FabricBitstreamMemoryBank here. Client functions should use the API from FabricBitstream only. If you can create wrapper APIs on FabricBitstream, it would be great.

Reply:
There is quite some data we want to access from sub data structure, to list a few: get_longest_effective_wl_count(), get_total_bl_addr_size(), get_total_wl_addr_size(), blwl_lengths, datas, masks and etc

If everything need to be access via wrapper, I will think it is inefficient and more importantly I would think it is ugly coding (personal preference)

So I do not plan to address this. But If you insist, then please file a case, we can address that in new PR.